HCIDOCS-623: Docs for OCPSTRAT-1819 [Tech Preview] BareMetal as a Service Support for OpenShift (#93601)
openshift/openshift-docs: modules/bmo-configuring-userdata.adoc

// This module is included in the following assemblies:
//
// * installing/installing_bare_metal/bare-metal-using-bare-metal-as-a-service.adoc

:_mod-docs-content-type: PROCEDURE
[id="bmo-configuring-userdata_{context}"]
= Configuring userData

The cloud image for most operating systems provides the `cloud-init` service, which can read the custom scripts that you create in the `userData` format.

> The `userData` field in the `BareMetalHost` custom resource allows you to customize the bare-metal host during provisioning by providing initial configuration data, such as user credentials or system settings. For most operating systems, this is achieved by using a `cloud-init` script encapsulated in a Kubernetes secret.

Reviewer: ```suggestion
Customize the bare-metal host during provisioning by providing initial configuration data in the form of a script by using the `userData` field in the `BareMetalHost` custom resource (CR). For example, you can provide user credentials and system settings. For most operating systems, this is achieved by using a `cloud-init` script encapsulated in a Kubernetes secret. The `cloud-init` service can read the custom scripts that you create in the `userData` format.
```
hm, are we using the field value of a parameter, or creating scripts that run as instructed by the field?

Create and apply the `userData` for operating systems that use `cloud-init`.

.Procedure

. Create a file named `user-data.yaml` with the following content:
+
[source,yaml]
----
users:

> - name: <name>

Reviewer: hyphen?

  sudo: ["ALL=(ALL) NOPASSWD:ALL"]
  ssh_authorized_keys:
  - <ssh_key>
----

> +

Reviewer: ```suggestion
where:
```

<name>::
    The user name.
<ssh_key>::
    The public SSH key to use when accessing this host as the `<name>` user.

. Create a secret from the `userData` file by running the following command:
+
[source,terminal]
----

> $ oc create secret generic <name>-user-data \

Reviewer: replaceable value definition; <name>?

  --from-file=userData=user-data.yaml -n bmaas
----

. Configure the `BareMetalHost` to use the `userData` file by running the following command:
+
[source,terminal]
----
$ oc patch baremetalhost <hostname> -n bmaas \
     --type merge -p '{"spec":{"userData":{"name":"<name>-user-data"}}}'
----

> +

Reviewer: ```suggestion
where:
```

// This module is included in the following assemblies:
//
// * installing/installing_bare_metal/bare-metal-using-bare-metal-as-a-service.adoc

:_mod-docs-content-type: PROCEDURE
[id="bmo-configuring-userdata_{context}"]
= Configuring userData

The cloud image for most operating systems provides the `cloud-init` service, which can read the custom scripts that you create in the `userData` format.

The `userData` field in the `BareMetalHost` custom resource allows you to customize the bare-metal host during provisioning by providing initial configuration data, such as user credentials or system settings. For most operating systems, this is achieved by using a `cloud-init` script encapsulated in a Kubernetes secret. 

> [NOTE]

Reviewer 2: remove for now

Author: Done.

====

> For {op-system-first}, use Ignition.

Reviewer 2: this comes out as 
"For Red Hat Enterprise Linux CoreOS (RHCOS), use Ignition."
but we don't support RHCOS on BMaaS (only Fedora/CentOS CoreOS)
let's revisit and make sure this is expanded in a way that is consistent with other OS references.

Author: Yeah. They have some conditions in the attributes file, so I just inlined the text.
